cmd/nvidia-dra-plugin/device_state.go

```diff
@@ -349,35 +357,70 @@ func (s *DeviceState) prepareDevices(ctx context.Context, claim *resourceapi.Res
 
 func (s *DeviceState) unprepareDevices(ctx context.Context, claimUID string, devices PreparedDevices) error {
 	for _, group := range devices {
-		// Stop any MPS control daemons started for each group of prepared devices.
-		mpsControlDaemon := s.mpsManager.NewMpsControlDaemon(claimUID, group)
-		if err := mpsControlDaemon.Stop(ctx); err != nil {
-			return fmt.Errorf("error stopping MPS control daemon: %w", err)
+		var err error
+		if group.ConfigState.GpuConfig != nil {
+			err = s.unprepareGpus(ctx, group.ConfigState.GpuConfig, group.Devices.Gpus())
 		}
-
-		// Go back to default time-slicing for all full GPUs.
-		tsc := configapi.DefaultGpuConfig().Sharing.TimeSlicingConfig
-		if err := s.tsManager.SetTimeSlice(group.Devices.Gpus(), tsc); err != nil {
-			return fmt.Errorf("error setting timeslice for devices: %w", err)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (s *DeviceState) unprepareGpus(ctx context.Context, config *configapi.GpuConfig, devices PreparedDeviceList) error {
+	if config.DriverConfig.Driver == configapi.VfioPciDriver {
+		for _, device := range devices {
+			if err := s.vfioPciManager.Unconfigure(device.Gpu.Info); err != nil {
+				return fmt.Errorf("error unconfiguring vfio-pci device: %w", err)
+			}
 		}
 	}
+	// Go back to default time-slicing for all full GPUs.
+	tsc := configapi.DefaultGpuConfig().Sharing.TimeSlicingConfig
+	if err := s.tsManager.SetTimeSlice(devices, tsc); err != nil {
+		return fmt.Errorf("error setting timeslice for devices: %w", err)
+	}
 	return nil
 }
 
 func (s *DeviceState) applyConfig(ctx context.Context, config configapi.Interface, claim *resourceapi.ResourceClaim, results []*resourceapi.DeviceRequestAllocationResult) (*DeviceConfigState, error) {
+	var err error
+	var configState DeviceConfigState
+
 	switch castConfig := config.(type) {
 	case *configapi.GpuConfig:
-		return s.applySharingConfig(ctx, castConfig.Sharing, claim, results)
+		configState.GpuConfig = castConfig
+		err = s.applyGpuConfig(ctx, castConfig, claim, results, &configState)
```

Comment:
Why not still return if `err != nil` in each of the branches? alternatively, does it make sense to have `s.applyGpuConfig` return `(*DeviceConfigState, error)` like the other apply functions and keep the other function unchanged?